go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,304 @@
+	nsInfo, err := oc.waitForNamespaceLocked(egressFirewall.Namespace)
+	if err != nil {
+		klog.Errorf("Failed to wait for namespace %s event (%v)",
+			egressFirewall.Namespace, err)
+		return
+	}
+	defer nsInfo.Unlock()
+
+	if nsInfo.egressFirewall {
+		klog.Errorf("Attempting to add egressFirewall %s to namespace %s when it already has an egressFirewall",
+			egressFirewall.Name, egressFirewall.Namespace)
+		return
+	}
+
+	ef := newEgressFirewall(egressFirewall)
+	nsInfo.egressFirewallPolicy = ef
+	//lock the newgressFirewall and unlock nsInfo
+	for i, egressFirewallRule := range egressFirewall.Spec.Egress {
+		//process Rules into egressFirewallRules for egressFirewall struct
+		efr, err := newEgressFirewallRule(egressFirewallRule, i)
+		if err != nil {
+			klog.Errorf("Cannot create EgressFirewall Rule for destination %s to namespace %s - %v",
+				egressFirewallRule.To.CIDRSelector, egressFirewall.Namespace, err)
+			continue
+
+		}
+		ef.egressRules = append(ef.egressRules, efr)
+	}
+
+	existingNodes, err := oc.kube.GetNodes()
+	if err != nil {
+		klog.Errorf("Unable to add egressfirewall %s, cannot list nodes: %s", egressFirewall.Name, err)
+	}
+	for _, node := range existingNodes.Items {
+		joinSwitch := joinSwitch(node.Name)
+		err = ef.addACLToJoinSwitch(joinSwitch, nsInfo.addressSet.GetIPv4HashName())
+		if err != nil {
+			klog.Errorf("%s", err)
+		}
+	}
+
+}
+
+func (oc *Controller) updateEgressFirewall(oldEgressFirewall, newEgressFirewall *egressfirewallapi.EgressFirewall) {
+	if !reflect.DeepEqual(oldEgressFirewall.Spec, newEgressFirewall.Spec) {
+		oc.deleteEgressFirewall(oldEgressFirewall)
+		oc.addEgressFirewall(newEgressFirewall)
+	}
+}
+
+func (oc *Controller) deleteEgressFirewall(egressFirewall *egressfirewallapi.EgressFirewall) {
+	klog.Infof("Deleting egress Firewall %s in namespace %s", egressFirewall.Name, egressFirewall.Namespace)
+
+	nsInfo, err := oc.waitForNamespaceLocked(egressFirewall.Namespace)
+	if err != nil {
+		klog.Errorf("Failed to wait for namespace %s event (%v)",
+			egressFirewall.Namespace, err)
+	}
+	defer nsInfo.Unlock()
+	nsInfo.egressFirewall = false
+
+	stdout, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=_uuid", "find", "ACL",
+		fmt.Sprintf("external-ids:egressFirewall=%s", egressFirewall.Namespace))
+	if err != nil {
+		klog.Errorf(" error executing create ACL command, stderr: %q, %+v", stderr, err)
+		//none of the elements of the egressfirewall have been deleted
+		nsInfo.egressFirewall = true
+		return
+	}
+	uuids := strings.Fields(stdout)
+	existingNodes, err := oc.kube.GetNodes()
+	if err != nil {
+		klog.Errorf("Error deleting egressFirewall for namespace %s, cannot get nodes to delete ACLS %v",
+			egressFirewall.Namespace, err)
+	}
+	for _, uuid := range uuids {
+		for _, node := range existingNodes.Items {
+			_, stderr, err := util.RunOVNNbctl("remove", "logical_switch",
```

Comment:
Remind me here: do we delete all ACL's when removing the `logical_switch`?

Reply:
@alexanderConstantinescu I don't understand what you mean. Here I am not removing a logical switch the command goes on to the next line `util.RunOVNNbctl("remove", "logical_switch",joinSwitch(node.Name), "acls", uuid)` I am removing from the logical switch the ACL with the given uuid. When all referances to an ACL have been removed from the database the ACL gets deleted because nothing is referencing it @dcbw am I right about this?

Reply:
@JacobTanenbaum I believe so, yes. Should probably check with the OVN team though.

Reply:
it does